src/core.jl

```diff
@@ -0,0 +1,170 @@
+"""
+Type representing any object drawable on image
+"""
+abstract Drawable
+
+"""
+    p = Point(x,y)
+    p = Point(c)
+
+A `Drawable` point on the image
+"""
+immutable Point <: Drawable
+    x::Int
+    y::Int
+end
+
+abstract Line <: Drawable
+abstract Circle <: Drawable
+
+
+"""
+    line = LineTwoPoints(p1, p2)
+
+A `Drawable` infinite length line passing through the two points
+`p1` and `p2`.
+"""
+immutable LineTwoPoints <: Line
+    p1::Point
+    p2::Point
+end
+
+"""
+    line = LineNormal(ρ, θ)
+
+A `Drawable` infinte length line having perpendicular length `ρ` from
+origin and angle `θ` between the perpendicular and x-axis
+"""
+immutable LineNormal <: Line
+    ρ::Real
+    θ::Real
```

Comment:
The performance difference you are observing is probably mostly due to the fact that the member variables of `LineNormal` need to be allocated on the heap. This is because those member variables are declared as abstract types. Consequently the compiler doesn't know beforehand how much space they need. 

Concerning your `@code_warntype` question: The parameters to the constructor are inferrable and so is the return type. That means there is no inference problem to report and `@code_warntype` is doing it's job.  However, once you have a function that uses the member variables, which are not inferable, then you'll see the red `Any` pop up as a consequence:

```julia
julia> immutable Point
           x::Real
       end

julia> f(p, i) = p.x * i
f (generic function with 1 method)

julia> p = Point(0.7)
Point(0.7)

julia> @code_warntype f(p, 0.7)
Variables:
  #self#::#f
  p::Point
  i::Float64

Body:
  begin 
      return ((Core.getfield)(p::Point,:x)::Real * i::Float64)::Any
  end::Any
```